zebrad/src/components/inbound.rs

```diff
@@ -115,85 +145,113 @@ impl Service<zn::Request> for Inbound {
+        //    and drops new requests after it reaches a hard-coded limit. This is the
+        //    "load shed directly" pattern from #1618.
+        //  * currently, the state service is always ready, unless its buffer is full.
+        //    So we might also want to propagate backpressure from its buffer.
+        //  * if we want to propagate backpressure, add a ReadyCache for each service, to ensure
+        //    that each poll_ready has a matching call. See #1593 for details.
+        Poll::Ready(Ok(()))
     }
 
     #[instrument(name = "inbound", skip(self, req))]
     fn call(&mut self, req: zn::Request) -> Self::Future {
         match req {
-            zn::Request::Peers => match self.address_book.as_ref() {
-                Some(addrs) => {
+            zn::Request::Peers => {
+                if let Setup::Initialized { address_book, .. } = &self.network {
                     // We could truncate the list to try to not reveal our entire
                     // peer set. But because we don't monitor repeated requests,
                     // this wouldn't actually achieve anything, because a crawler
                     // could just repeatedly query it.
-                    let mut peers = addrs.lock().unwrap().sanitized();
+                    let mut peers = address_book.lock().unwrap().sanitized();
                     const MAX_ADDR: usize = 1000; // bitcoin protocol constant
                     peers.truncate(MAX_ADDR);
                     async { Ok(zn::Response::Peers(peers)) }.boxed()
+                } else {
+                    info!("ignoring `Peers` request from remote peer during network setup");
+                    async { Ok(zn::Response::Nil) }.boxed()
                 }
-                None => async { Err("not ready to serve addresses".into()) }.boxed(),
-            },
+            }
             zn::Request::BlocksByHash(hashes) => {
-                let state = self.state.clone();
-                let requests = futures::stream::iter(
-                    hashes
-                        .into_iter()
-                        .map(|hash| zs::Request::Block(hash.into())),
-                );
-
-                state
-                    .call_all(requests)
-                    .try_filter_map(|rsp| {
-                        futures::future::ready(match rsp {
-                            zs::Response::Block(Some(block)) => Ok(Some(block)),
+                // Correctness:
+                //
+                // We can't use `call_all` here, because it leaks buffer slots:
+                // https://github.com/tower-rs/tower/blob/master/tower/src/util/call_all/common.rs#L112
+                use futures::stream::TryStreamExt;
+                hashes
+                    .into_iter()
+                    .map(|hash| zs::Request::Block(hash.into()))
+                    .map(|request| self.state.clone().oneshot(request))
+                    .collect::<futures::stream::FuturesOrdered<_>>()
```

Comment:
I'm not sure if we are required to preserve the original request order in our response.

But we don't gain much performance or simplicity from an unordered response, and it's possible some nodes might expect order, so we might as well keep it.

Reply:
yea I wasn't sure either, at least this way its pretty trivial to swap it to unordered if we find out later we can go faster safely